README.md
<!-- @format -->

This project was bootstrapped with [Create React Native App](https://github.com/react-community/create-react-native-app).

Below you'll find information about performing common tasks. The most recent version of this guide is available [here](https://github.com/react-community/create-react-native-app/blob/master/react-native-scripts/template/README.md).

## Table of Contents

-   [Updating to New Releases](#updating-to-new-releases)
-   [Available Scripts](#available-scripts)
    -   [npm start](#npm-start)
    -   [npm test](#npm-test)
    -   [npm run ios](#npm-run-ios)
    -   [npm run android](#npm-run-android)
    -   [npm run eject](#npm-run-eject)
-   [Writing and Running Tests](#writing-and-running-tests)
-   [Environment Variables](#environment-variables)
    -   [Configuring Packager IP Address](#configuring-packager-ip-address)
-   [Adding Flow](#adding-flow)
-   [Customizing App Display Name and Icon](#customizing-app-display-name-and-icon)
-   [Sharing and Deployment](#sharing-and-deployment)
    -   [Publishing to Expo's React Native Community](#publishing-to-expos-react-native-community)
    -   [Building an Expo "standalone" app](#building-an-expo-standalone-app)
    -   [Ejecting from Create React Native App](#ejecting-from-create-react-native-app)
        -   [Build Dependencies (Xcode & Android Studio)](#build-dependencies-xcode-android-studio)
        -   [Should I Use ExpoKit?](#should-i-use-expokit)
-   [Troubleshooting](#troubleshooting)
    -   [Networking](#networking)
    -   [iOS Simulator won't open](#ios-simulator-wont-open)
    -   [QR Code does not scan](#qr-code-does-not-scan)

## Updating to New Releases

You should only need to update the global installation of `create-react-native-app` very rarely, ideally never.

Updating the `react-native-scripts` dependency of your app should be as simple as bumping the version number in `package.json` and reinstalling your project's dependencies.

Upgrading to a new version of React Native requires updating the `react-native`, `react`, and `expo` package versions, and setting the correct `sdkVersion` in `app.json`. See the [versioning guide](https://github.com/react-community/create-react-native-app/blob/master/VERSIONS.md) for up-to-date information about package version compatibility.

## Available Scripts

If Yarn was installed when the project was initialized, then dependencies will have been installed via Yarn, and you should probably use it to run these commands as well. Unlike dependency installation, command running syntax is identical for Yarn and NPM at the time of this writing.

### `npm start`

Runs your app in development mode.

Open it in the [Expo app](https://expo.io) on your phone to view it. It will reload if you save edits to your files, and you will see build errors and logs in the terminal.

Sometimes you may need to reset or clear the React Native packager's cache. To do so, you can pass the `--reset-cache` flag to the start script:

```
npm start -- --reset-cache
# or
yarn start -- --reset-cache
```

#### `npm test`

Runs the [jest](https://github.com/facebook/jest) test runner on your tests.

#### `npm run ios`

Like `npm start`, but also attempts to open your app in the iOS Simulator if you're on a Mac and have it installed.

#### `npm run android`

Like `npm start`, but also attempts to open your app on a connected Android device or emulator. Requires an installation of Android build tools (see [React Native docs](https://facebook.github.io/react-native/docs/getting-started.html) for detailed setup). We also recommend installing Genymotion as your Android emulator. Once you've finished setting up the native build environment, there are two options for making the right copy of `adb` available to Create React Native App:

##### Using Android Studio's `adb`

1. Make sure that you can run adb from your terminal.
2. Open Genymotion and navigate to `Settings -> ADB`. Select “Use custom Android SDK tools” and update with your [Android SDK directory](https://stackoverflow.com/questions/25176594/android-sdk-location).

##### Using Genymotion's `adb`

1. Find Genymotion’s copy of adb. On macOS for example, this is normally `/Applications/Genymotion.app/Contents/MacOS/tools/`.
2. Add the Genymotion tools directory to your path (instructions for [Mac](http://osxdaily.com/2014/08/14/add-new-path-to-path-command-line/), [Linux](http://www.computerhope.com/issues/ch001647.htm), and [Windows](https://www.howtogeek.com/118594/how-to-edit-your-system-path-for-easy-command-line-access/)).
3. Make sure that you can run adb from your terminal.

#### `npm run eject`

This will start the process of "ejecting" from Create React Native App's build scripts. You'll be asked a couple of questions about how you'd like to build your project.

**Warning:** Running eject is a permanent action (aside from whatever version control system you use). An ejected app will require you to have an [Xcode and/or Android Studio environment](https://facebook.github.io/react-native/docs/getting-started.html) set up.

## Customizing App Display Name and Icon

You can edit `app.json` to include [configuration keys](https://docs.expo.io/versions/latest/guides/configuration.html) under the `expo` key.

To change your app's display name, set the `expo.name` key in `app.json` to an appropriate string.

To set an app icon, set the `expo.icon` key in `app.json` to be either a local path or a URL. It's recommended that you use a 512x512 png file with transparency.

## Writing and Running Tests

This project is set up to use [jest](https://facebook.github.io/jest/) for tests. You can configure whatever testing strategy you like, but jest works out of the box. Create test files in directories called `__tests__` or with the `.test` extension to have the files loaded by jest. See the [the template project](https://github.com/react-community/create-react-native-app/blob/master/react-native-scripts/template/App.test.js) for an example test. The [jest documentation](https://facebook.github.io/jest/docs/en/getting-started.html) is also a wonderful resource, as is the [React Native testing tutorial](https://facebook.github.io/jest/docs/en/tutorial-react-native.html).

## Environment Variables

You can configure some of Create React Native App's behavior using environment variables.

### Configuring Packager IP Address

When starting your project, you'll see something like this for your project URL:

```
exp://192.168.0.2:19000
```

The "manifest" at that URL tells the Expo app how to retrieve and load your app's JavaScript bundle, so even if you load it in the app via a URL like `exp://localhost:19000`, the Expo client app will still try to retrieve your app at the IP address that the start script provides.

In some cases, this is less than ideal. This might be the case if you need to run your project inside of a virtual machine and you have to access the packager via a different IP address than the one which prints by default. In order to override the IP address or hostname that is detected by Create React Native App, you can specify your own hostname via the `REACT_NATIVE_PACKAGER_HOSTNAME` environment variable:

Mac and Linux:

```
REACT_NATIVE_PACKAGER_HOSTNAME='my-custom-ip-address-or-hostname' npm start
```

Windows:

```
set REACT_NATIVE_PACKAGER_HOSTNAME='my-custom-ip-address-or-hostname'
npm start
```

The above example would cause the development server to listen on `exp://my-custom-ip-address-or-hostname:19000`.

## Adding Flow

Flow is a static type checker that helps you write code with fewer bugs. Check out this [introduction to using static types in JavaScript](https://medium.com/@preethikasireddy/why-use-static-types-in-javascript-part-1-8382da1e0adb) if you are new to this concept.

React Native works with [Flow](http://flowtype.org/) out of the box, as long as your Flow version matches the one used in the version of React Native.

To add a local dependency to the correct Flow version to a Create React Native App project, follow these steps:

1. Find the Flow `[version]` at the bottom of the included [.flowconfig](.flowconfig)
2. Run `npm install --save-dev flow-bin@x.y.z` (or `yarn add --dev flow-bin@x.y.z`), where `x.y.z` is the .flowconfig version number.
3. Add `"flow": "flow"` to the `scripts` section of your `package.json`.
4. Add `// @flow` to any files you want to type check (for example, to `App.js`).

Now you can run `npm run flow` (or `yarn flow`) to check the files for type errors.
You can optionally use a [plugin for your IDE or editor](https://flow.org/en/docs/editors/) for a better integrated experience.

To learn more about Flow, check out [its documentation](https://flow.org/).

## Sharing and Deployment

Create React Native App does a lot of work to make app setup and development simple and straightforward, but it's very difficult to do the same for deploying to Apple's App Store or Google's Play Store without relying on a hosted service.

### Publishing to Expo's React Native Community

Expo provides free hosting for the JS-only apps created by CRNA, allowing you to share your app through the Expo client app. This requires registration for an Expo account.

Install the `exp` command-line tool, and run the publish command:

```
$ npm i -g exp
$ exp publish
```

### Building an Expo "standalone" app

You can also use a service like [Expo's standalone builds](https://docs.expo.io/versions/latest/guides/building-standalone-apps.html) if you want to get an IPA/APK for distribution without having to build the native code yourself.

### Ejecting from Create React Native App

If you want to build and deploy your app yourself, you'll need to eject from CRNA and use Xcode and Android Studio.

This is usually as simple as running `npm run eject` in your project, which will walk you through the process. Make sure to install `react-native-cli` and follow the [native code getting started guide for React Native](https://facebook.github.io/react-native/docs/getting-started.html).

#### Should I Use ExpoKit?

If you have made use of Expo APIs while working on your project, then those API calls will stop working if you eject to a regular React Native project. If you want to continue using those APIs, you can eject to "React Native + ExpoKit" which will still allow you to build your own native code and continue using the Expo APIs. See the [ejecting guide](https://github.com/react-community/create-react-native-app/blob/master/EJECTING.md) for more details about this option.

## Troubleshooting

### Networking

If you're unable to load your app on your phone due to a network timeout or a refused connection, a good first step is to verify that your phone and computer are on the same network and that they can reach each other. Create React Native App needs access to ports 19000 and 19001 so ensure that your network and firewall settings allow access from your device to your computer on both of these ports.

Try opening a web browser on your phone and opening the URL that the packager script prints, replacing `exp://` with `http://`. So, for example, if underneath the QR code in your terminal you see:

```
exp://192.168.0.1:19000
```

Try opening Safari or Chrome on your phone and loading

```
http://192.168.0.1:19000
```

and

```
http://192.168.0.1:19001
```

If this works, but you're still unable to load your app by scanning the QR code, please open an issue on the [Create React Native App repository](https://github.com/react-community/create-react-native-app) with details about these steps and any other error messages you may have received.

If you're not able to load the `http` URL in your phone's web browser, try using the tethering/mobile hotspot feature on your phone (beware of data usage, though), connecting your computer to that WiFi network, and restarting the packager.

### iOS Simulator won't open

If you're on a Mac, there are a few errors that users sometimes see when attempting to `npm run ios`:

-   "non-zero exit code: 107"
-   "You may need to install Xcode" but it is already installed
-   and others

There are a few steps you may want to take to troubleshoot these kinds of errors:

1. Make sure Xcode is installed and open it to accept the license agreement if it prompts you. You can install it from the Mac App Store.
2. Open Xcode's Preferences, the Locations tab, and make sure that the `Command Line Tools` menu option is set to something. Sometimes when the CLI tools are first installed by Homebrew this option is left blank, which can prevent Apple utilities from finding the simulator. Make sure to re-run `npm/yarn run ios` after doing so.
3. If that doesn't work, open the Simulator, and under the app menu select `Reset Contents and Settings...`. After that has finished, quit the Simulator, and re-run `npm/yarn run ios`.

### QR Code does not scan

If you're not able to scan the QR code, make sure your phone's camera is focusing correctly, and also make sure that the contrast on the two colors in your terminal is high enough. For example, WebStorm's default themes may [not have enough contrast](https://github.com/react-community/create-react-native-app/issues/49) for terminal QR codes to be scannable with the system barcode scanners that the Expo app uses.

If this causes problems for you, you may want to try changing your terminal's color theme to have more contrast, or running Create React Native App from a different terminal. You can also manually enter the URL printed by the packager script in the Expo app's search bar to load it manually.


CHANGELOG.md
<!-- @format -->

# CHANGE LOG

<!-- Format:
## 23 Jan 2018, by Jia Zeng <andyzeng96@gmail.com>
* Change 1, issue #12
* Change 2
* Change 3

if there is corresponding issue, state it
-->

-   Version 0.0.1

## 4 June 2018, by Jia Zeng <andyzeng96@gmail.com>

-   logo image on splash, tappable footer
-   registration buttons tappable and opacity
-   lighter gray for profile summary page divider
-   Profile edit form profile image styling
-   Setting divider lighter gray
-   fix multiline input

## 3 June 2018, by Jia Zeng <andyzeng96@gmail.com>

-   Test update friendship in profile page (block, unblock, request, withdraw request, accept and rejects)
-   action button design change for profile page (darken gray background)
-   Bug fix for updating reducer in request friendship in profile page

## 1 June 2018, by Jia Zeng <andyzeng96@gmail.com>

-   Refactor to create factory function for ActionSheetIOS
-   Update SearchBarHeader to include Profile Friendship setting
    that includes block and report user
-   Bug fix for rendering # of mutual friends
-   Update profile detail design
    -   Occupation
    -   divider
    -   Elevator pitch
    -   about

## 1 June 2018, by Jia Zeng <andyzeng96@gmail.com>

-   Empty results for suggested (“No Recommendations ”)
-   remove friends … button
-   change profileSummaryCard button to blue background circle >
-   Empty results for search results (No Results)
-   change search card button to blue background circle >
-   A lot of refactoring

## 30 Apr 2018, by Jia Zeng <andyzeng96@gmail.com>

-   Center search text

## 29 Apr 2018, by Jia Zeng <andyzeng96@gmail.com>

-   Initial implementation for MeetTab and MeetCard UI done
-   Integrate phone verification status code passed back from server

## 27 Apr 2018, by Jia Zeng <andyzeng96@gmail.com>

-   Refactor and add redux persist to persist reducers
-   Image phone verification logic with deep link

## 26 Apr 2018, by Jia Zeng <andyzeng96@gmail.com>

-   Update image resizing size
-   Create password update UI and hook up with data

## 24 Apr 2018, by Jia Zeng <andyzeng96@gmail.com>

-   Update splash screen fonts and image
-   Refactor login page and account creation page

## 23 Apr 2018, by Jia Zeng <andyzeng96@gmail.com>

-   See [Issue #33](https://github.com/somechild/goalmogul-ios/issues/33)
-   Create Splash screen
-   Implement splash screen actions
-   Update friends privacy setting tick icon

## 22 Apr 2018, by Jia Zeng <andyzeng96@gmail.com>

-   See [Issue #30](https://github.com/somechild/goalmogul-ios/issues/30)
-   See [Issue #31](https://github.com/somechild/goalmogul-ios/issues/31)
-   Create setting page UI + hook up with data
-   Create email managing page UI
-   Create email update page UI + hook up with data
-   Create phone managing page UI
-   Create add phone number page UI
-   Create privacy friends setting page UI

## 22 Apr 2018, by Jia Zeng <andyzeng96@gmail.com>

-   See [Issue #30](https://github.com/somechild/goalmogul-ios/issues/30)
-   Create setting page

## 22 Apr 2018, by Jia Zeng <andyzeng96@gmail.com>

-   See [Issue #34](https://github.com/somechild/goalmogul-ios/issues/34)
-   reorder inputs
-   length limitation

## 16 Apr 2018, by Jia Zeng <andyzeng96@gmail.com>

-   Implement profile editing form UI
-   Connect profile edit form UI with data

## 12 Apr 2018, by Jia Zeng <andyzeng96@gmail.com>

-   Bug fix image uploading
-   Conditional rendering for edit button
-   Hook up profile cards with real data

## 2 Apr 2018, by Jia Zeng <andyzeng96@gmail.com>

-   Finish Profile page UI (Apr 2nd)
-   Profile detail card UI
-   Link button actions among profile pages
-   Image uploading

## 25 Mar 2018, by Jia Zeng <andyzeng96@gmail.com>

-   Add ghost image for profile picture
-   Disable InputText onFocus enlarge
-   Login and Registration account error handling
-   Link registration and login workflow with server
-   Set up initial ui for profile and profile detail
-   Implement scrollable view on keyboard showing up
-   Refactor main app navigations from TabBarIOS to react native router flux

## 25 Mar 2018, by Jia Zeng <andyzeng96@gmail.com>

-   See [Issue #17](https://github.com/somechild/goalmogul-ios/issues/17)
-   Registration error handling
-   Registration skip button implementation
-   Form redesign

## 22 Mar 2018, by Jia Zeng <andyzeng96@gmail.com>

-   See [Issue #11](https://github.com/somechild/goalmogul-ios/issues/11)
-   Registration import photo for profile picture

## 22 Feb 2018, by Jia Zeng <andyzeng96@gmail.com>

-   See [Issue #7](https://github.com/somechild/goalmogul-ios/issues/7)
-   Set up routing between pages for registration
-   Set up react-native-router-flux package
-   Implement basic Registration UI

## 22 Feb 2018, by Jia Zeng <andyzeng96@gmail.com>

-   See [Issue #9](https://github.com/somechild/goalmogul-ios/issues/9)
-   Implement header pagination


src/SplashScreen.js
/** @format */

import { AppLoading } from 'expo'
import React, { Component } from 'react'
import { Dimensions, Image, Text, View } from 'react-native'
import { Actions } from 'react-native-router-flux'
import { connect } from 'react-redux'
// Actions
import { tryAutoLoginV2, loadInitialAssets } from './actions'
/* Asset */
import LogoText from './asset/header/GMText.png'
import LogoIcon from './asset/header/header-logo-white.png'
import HandshakeArt from './asset/header/handshake.png'
import { trackViewScreen } from './monitoring/segment'
import { Screen } from './monitoring/segment/Constants'
import { IS_BIG_IPHONE } from './Utils/Constants'
// Components
import { color, default_style } from './styles/basic'
import { FONT_FAMILY, LETTER_SPACING } from './styles/basic/text'
import { UI_SCALE } from './styles'
import DelayedButton from './Main/Common/Button/DelayedButton'
import {
    getBottomSpace,
    getStatusBarHeight,
} from 'react-native-iphone-x-helper'

const WINDOW_WIDTH = Dimensions.get('window').width
const DEBUG_KEY = '[ UI SplashScreen ]'

class SplashScreen extends Component {
    constructor(props) {
        super(props)
        this.state = {
            fontLoaded: false,
            appReady: false,
        }
    }

    async componentDidMount() {
        this.setState({ fontLoaded: true })
        trackViewScreen(Screen.SPLASH_SCREEN)
    }

    handleGetStartedOnPress() {
        this.props.registration()
    }

    handleLoginPress() {
        this.props.login()
    }

    render() {
        if (!this.state.appReady) {
            return (
                <AppLoading
                    startAsync={async () => {
                        // Load necessary assets like fonts
                        await loadInitialAssets()

                        // Attempt auto login
                        await this.props.tryAutoLoginV2()
                    }}
                    onFinish={() => this.setState({ appReady: true })}
                    onError={console.warn}
                    autoHideSplash={false}
                />
            )
        }

        return (
            <View style={styles.containerStyle}>
                {/* Main content including logo, artwork and slogan */}
                <View style={styles.bodyContainerStyle}>
                    {/* Top header logo */}
                    <View style={styles.logoContainerStyle}>
                        <Image source={LogoIcon} style={styles.logoIconStyle} />
                        <Image source={LogoText} style={styles.logoTextStyle} />
                    </View>
                    {/* Artwork */}
                    <Image source={HandshakeArt} style={styles.artworkStyle} />
                    {/* Slogan */}
                    <Text style={styles.sloganTextStyle}>
                        Achieve more, together.
                    </Text>
                </View>
                {/* Bottom bar for registration and log in */}
                <View style={styles.actionButtonsContainerStyle}>
                    <DelayedButton
                        onPress={this.handleGetStartedOnPress.bind(this)}
                        touchableWithoutFeedback
                    >
                        <View style={styles.signUpButtonStyle}>
                            <Text
                                style={[
                                    default_style.buttonText_1,
                                    {
                                        color: color.GM_BLUE_DEEP,
                                    },
                                ]}
                            >
                                Sign Up
                            </Text>
                        </View>
                    </DelayedButton>
                    <DelayedButton
                        onPress={this.handleLoginPress.bind(this)}
                        touchableWithoutFeedback
                    >
                        <Text
                            style={[
                                default_style.buttonText_1,
                                {
                                    color: 'white',
                                    paddingVertical: 16,
                                },
                            ]}
                        >
                            Log In
                        </Text>
                    </DelayedButton>
                </View>
            </View>
        )
    }
}

const styles = {
    // containers
    containerStyle: {
        flex: 1,
        backgroundColor: color.GM_BLUE_DEEP,
        paddingBottom: getBottomSpace(),
        paddingTop: getStatusBarHeight(true),
    },
    bodyContainerStyle: {
        flexGrow: 1,
        alignItems: 'center',
        justifyContent: 'center',
    },
    actionButtonsContainerStyle: {
        paddingVertical: 24,
        alignItems: 'center',
        justifyContent: 'center',
        paddingHorizontal: 16,
    },

    // Logo
    logoContainerStyle: {
        justifyContent: 'center',
        alignItems: 'center',
        marginBottom: IS_BIG_IPHONE ? 60 : 48,
    },
    logoIconStyle: {
        height: 56,
        width: 56,
        marginBottom: 16,
        resizeMode: 'contain',
    },
    logoTextStyle: {
        height: 36,
        width: 208,
        resizeMode: 'contain',
    },

    // center artwork and slogan
    artworkStyle: {
        height: IS_BIG_IPHONE ? 180 : 160,
        width: WINDOW_WIDTH,
        resizeMode: 'contain',
        marginVertical: 24,
        tintColor: 'white',
    },
    sloganTextStyle: {
        color: 'white',
        fontFamily: FONT_FAMILY.MEDIUM,
        fontSize: 24 * UI_SCALE,
        letterSpacing: LETTER_SPACING.WIDE,
    },

    // action button styles
    signUpButtonStyle: {
        paddingVertical: 12,
        backgroundColor: 'white',
        width: '100%',
        borderRadius: 3,
        alignItems: 'center',
    },
}

const mapDispatchToProps = (dispatch) => {
    return {
        registration: () => Actions.push('registrationAccount'),
        login: () => Actions.push('login'),
        tryAutoLoginV2: (params) => dispatch(tryAutoLoginV2(params)),
    }
}

export default connect(null, mapDispatchToProps)(SplashScreen)
